BUGFIX: Resolve issue with custom data source displaying additional l…

markusguenther opened the pull request:
Previously, there was an issue with the custom data source where the secondary label, utilized with label and preview, was not being displayed correctly. This commit addresses the problem, ensuring proper alignment and display of the secondary label in conjunction with the label and preview elements. The implementation includes the use of a CSS grid layout to achieve consistent and accurate rendering of the custom data source information.

Fixes: #3675

**Before:**
<img width="397" alt="Screenshot 2024-01-26 at 13 28 55" src="https://github.com/neos/neos-ui/assets/1014126/595550e1-6652-4b48-8930-b4a6ba217637">

**After with preview:**
<img width="335" alt="Screenshot 2024-01-26 at 14 13 09" src="https://github.com/neos/neos-ui/assets/1014126/bf0c6560-0d9e-499f-86c8-97e35cc6fb45">

**After without preview and 3rd label:**
<img width="325" alt="Screenshot 2024-01-26 at 14 17 11" src="https://github.com/neos/neos-ui/assets/1014126/5d24238e-c30a-40a6-99e0-be8aed50fb28">

markusguenther committed:
BUGFIX: Resolve issue with custom data source displaying additional labels incorrectly

Previously, there was an issue with the custom data source where the secondary label, utilized with label and preview, was not being displayed correctly. This commit addresses the problem, ensuring proper alignment and display of the secondary label in conjunction with the label and preview elements. The implementation includes the use of a CSS grid layout to achieve consistent and accurate rendering of the custom data source information.

Fixes: #3675

markusguenther commented:
For easier testing here my DataSource:

```php
<?php

namespace Neos\Demo\Service\DataSource;


use Neos\Flow\Annotations as Flow;
use Neos\Flow\Persistence\PersistenceManagerInterface;
use Neos\Neos\Domain\Service\UserService;
use Neos\Neos\Service\DataSource\AbstractDataSource;
use Neos\ContentRepository\Domain\Model\NodeInterface;

class EditorsDataSource extends AbstractDataSource
{

    /**
     * @var string
     */
    static protected $identifier = 'neos-demo-editors';

    /**
     * @Flow\Inject
     * @var UserService
     */
    protected $userService;

    /**
     * @Flow\Inject
     * @var PersistenceManagerInterface
     */
    protected $persistenceManager;

    /**
     * @param NodeInterface $node The node that is currently edited (optional)
     * @param array $arguments Additional arguments (key / value)
     * @return array
     */
    public function getData(NodeInterface $node = null, array $arguments = [])
    {
        $options = [];
        foreach ($this->userService->getUsers() as $user) {
			$options[] = [
                'value' => $this->persistenceManager->getIdentifierByObject($user),
                'label' => $user->getLabel(),

                // additional optional parameters:
                'secondaryLabel' => "works here",
                'tertiaryLabel' => "is a nice person",
                'preview' => 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHZpZXdCb3g9JzAgMCA0NDggNTEyJz48IS0tIUZvbnQgQXdlc29tZSBGcmVlIDYuNS4xIGJ5IEBmb250YXdlc29tZSAtIGh0dHBzOi8vZm9udGF3ZXNvbWUuY29tIExpY2Vuc2UgLSBodHRwczovL2ZvbnRhd2Vzb21lLmNvbS9saWNlbnNlL2ZyZWUgQ29weXJpZ2h0IDIwMjQgRm9udGljb25zLCBJbmMuLS0+PHBhdGggZD0nTTIyNCAxNmMtNi43IDAtMTAuOC0yLjgtMTUuNS02LjFDMjAxLjkgNS40IDE5NCAwIDE3NiAwYy0zMC41IDAtNTIgNDMuNy02NiA4OS40QzYyLjcgOTguMSAzMiAxMTIuMiAzMiAxMjhjMCAxNC4zIDI1IDI3LjEgNjQuNiAzNS45Yy0uNCA0LS42IDgtLjYgMTIuMWMwIDE3IDMuMyAzMy4yIDkuMyA0OEg0NS40QzM4IDIyNCAzMiAyMzAgMzIgMjM3LjRjMCAxLjcgLjMgMy40IDEgNWwzOC44IDk2LjlDMjguMiAzNzEuOCAwIDQyMy44IDAgNDgyLjNDMCA0OTguNyAxMy4zIDUxMiAyOS43IDUxMkg0MTguM2MxNi40IDAgMjkuNy0xMy4zIDI5LjctMjkuN2MwLTU4LjUtMjguMi0xMTAuNC03MS43LTE0M0w0MTUgMjQyLjRjLjYtMS42IDEtMy4zIDEtNWMwLTcuNC02LTEzLjQtMTMuNC0xMy40SDM0Mi43YzYtMTQuOCA5LjMtMzEgOS4zLTQ4YzAtNC4xLS4yLTguMS0uNi0xMi4xQzM5MSAxNTUuMSA0MTYgMTQyLjMgNDE2IDEyOGMwLTE1LjgtMzAuNy0yOS45LTc4LTM4LjZDMzI0IDQzLjcgMzAyLjUgMCAyNzIgMGMtMTggMC0yNS45IDUuNC0zMi41IDkuOWMtNC44IDMuMy04LjggNi4xLTE1LjUgNi4xem01NiAyMDhIMjY3LjZjLTE2LjUgMC0zMS4xLTEwLjYtMzYuMy0yNi4yYy0yLjMtNy0xMi4yLTctMTQuNSAwYy01LjIgMTUuNi0xOS45IDI2LjItMzYuMyAyNi4ySDE2OGMtMjIuMSAwLTQwLTE3LjktNDAtNDBWMTY5LjZjMjguMiA0LjEgNjEgNi40IDk2IDYuNHM2Ny44LTIuMyA5Ni02LjRWMTg0YzAgMjIuMS0xNy45IDQwLTQwIDQwem0tODggOTZsMTYgMzJMMTc2IDQ4MCAxMjggMjg4bDY0IDMyem0xMjgtMzJMMjcyIDQ4MCAyNDAgMzUybDE2LTMyIDY0LTMyeicvPjwvc3ZnPg=='
            ];
        }
        return $options;
    }
}

```

And the nodetype stuff:
```yaml
properties:
    authors:
      type: array
      ui:
        label: 'Author(s)'
        reloadIfChanged: true
        inspector:
          group: 'document'
          position: '200'
          editor: Neos.Neos/Inspector/Editors/SelectBoxEditor
          editorOptions:
            placeholder: Choose
            dataSourceIdentifier: neos-demo-editors
```

Sebobo commented on packages/react-ui-components/src/SelectBox_Option_MultiLineWithThumbnail/style.css:
```diff
@@ -1,8 +1,25 @@
     background-color: var(--colors-ContrastDarkest);
 }
 
+.multiLineWithThumbnail__item span {
```
A looong tittle makes the whole element wider and the ellipsis of the secondary and tertiary labels disappear.

Sebobo commented on packages/react-ui-components/src/SelectBox_Option_MultiLineWithThumbnail/style.css:
```diff
@@ -1,8 +1,25 @@
 .multiLineWithThumbnail__item {
+    display: grid;
+    grid-template-columns: auto 1fr;
+    gap: -var(--spacing-Half);
```
A negative gap is marked as invalid value and also doesn't seem to have any effect.

The styling would also be simpler when the gap is actually used and the padding of the element disabled and therefore the image correctly positioned (also the image margin needs to be disabled.

The result is looking the same but without all the CSS pushing around.

Sebobo reviewed (changes requested):
Thx, looks better (though I don't like the image hanging loose there, which is not a problem of this PR), but this change needs a cleaner CSS implementation.

markusguenther committed:
BUGFIX: Adjust styles for long labels

Long labels float out of the container and do not look nice. This adds an ellipsis and the tooltip is always available on hover to read the whole text. The Image is now also aligned in centered position.

markusguenther commented:
Thanks for the valuable feedback. With very long labels, it looked like that:
<img width="305" alt="Screenshot 2024-02-02 at 13 35 23" src="https://github.com/neos/neos-ui/assets/1014126/7abfb9b8-b8d3-4978-bcdd-34538b3196bc">
<img width="305" alt="Screenshot 2024-02-02 at 13 35 12" src="https://github.com/neos/neos-ui/assets/1014126/d9e5dfa8-2ead-4f38-a785-270a928f0231">

So the idea was to break the text, but it leads to huge heights.
<img width="306" alt="Screenshot 2024-02-02 at 13 31 58" src="https://github.com/neos/neos-ui/assets/1014126/7d076955-f90d-47bc-ab0c-da17f89924af">

Now we add an ellipsis and, like in the `NodeInfoView` we have a tooltip by default.

**Safari:**
<img width="432" alt="Screenshot 2024-02-02 at 14 18 28" src="https://github.com/neos/neos-ui/assets/1014126/7d17f93d-f70f-4c40-b1c5-6b60b7fea984">
**Chrome:**
<img width="464" alt="Screenshot 2024-02-02 at 14 17 30" src="https://github.com/neos/neos-ui/assets/1014126/4b4d294f-bfba-4006-8e02-7a6d3c2701cd">
**Firefox:**
<img width="337" alt="Screenshot 2024-02-02 at 14 17 18" src="https://github.com/neos/neos-ui/assets/1014126/3da7449a-305c-4c06-a393-8302a1b8aa3c">

Sebobo commented on packages/react-ui-components/src/SelectBox_Option_MultiLineWithThumbnail/style.css:
```diff
@@ -26,17 +46,17 @@
-    margin-left: -var(--spacing-Full);
-    margin-top: -var(--spacing-Half);
-    margin-bottom: -var(--spacing-Half);
+    margin-left: -var(--spacing-Half);
```
Not so happy about still having a negative margin, but I think its fine for now.

markusguenther replied:
Yeah, but in the select boxes I am always afraid that I break something else 🙈

Sebobo reviewed (approved):
Great much better!
